presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantAggregateDistinct.java

```diff
@@ -60,18 +57,36 @@ public Pattern<AggregationNode> getPattern()
     @Override
     public Result apply(AggregationNode node, Captures captures, Context context)
     {
+        ImmutableMap.Builder<VariableReferenceExpression, AggregationNode.Aggregation> aggregations = ImmutableMap.builder();
+        boolean ruleApplicable = false;
+
+        for (Map.Entry<VariableReferenceExpression, AggregationNode.Aggregation> agg : node.getAggregations().entrySet()) {
+            if (agg.getValue().isDistinct()) {
+                Set<VariableReferenceExpression> varAndGroupingKeySet =
+                        Stream.concat(node.getGroupingKeys().stream().map(VariableReferenceExpression.class::cast),
+                                        (agg.getValue()).getArguments().stream().map(VariableReferenceExpression.class::cast))
+                                .collect(Collectors.toSet());
+                if (((GroupReference) node.getSource()).getLogicalProperties().get().isDistinct(varAndGroupingKeySet)) {
+                    aggregations.put(agg.getKey(), removeDistinct(agg.getValue()));
+                    ruleApplicable = true;
+                }
+                else {
+                    aggregations.put(agg);
+                }
+            }
+        }
+
+        if (!ruleApplicable) {
```

Comment:
instead of this extra variable, can you build the aggregations map and check if it's equal to the original or not?